caffe2/operators/max_pool_with_index.cu

```diff
@@ -0,0 +1,197 @@
+    // find out the local index
+    // find out the local offset
+    const int w = index % width;
+    const int h = (index / width) % height;
+    const int c = (index / width / height) % channels;
+    const int n = index / width / height / channels;
+    const int phstart =
+         (h + pad_h < kernel_h) ? 0 : (h + pad_h - kernel_h) / stride_h + 1;
+    const int phend = min((h + pad_h) / stride_h + 1, pooled_height);
+    const int pwstart =
+         (w + pad_w < kernel_w) ? 0 : (w + pad_w - kernel_w) / stride_w + 1;
+    const int pwend = min((w + pad_w) / stride_w + 1, pooled_width);
+    float gradient = 0;
+    const int offset = (n * channels + c) * pooled_height * pooled_width;
+    const Dtype* const top_diff_slice = top_diff + offset;
+    const int* const mask_slice = mask + offset;
+
+    for (int ph = phstart; ph < phend; ++ph) {
+      for (int pw = pwstart; pw < pwend; ++pw) {
+        if (mask_slice[ph * pooled_width + pw] == h * width + w) {
+          gradient += convert::To<Dtype, float>(top_diff_slice[ph * pooled_width + pw]);
+        }
+      }
+    }
+    bottom_diff[index] = convert::To<float, Dtype>(gradient);
+  }
+}
+};
+
+
+template <typename T>
+bool MaxPoolWithIndexOp::DoRunWithType() {
+	auto& X = Input(0);
+  auto* Y = Output(0);
+  auto* mask = Output(1);
+
+  ConvPoolOpBase<CUDAContext>::SetOutputSize(X, Y, X.dim32(1));
+  int output_size = Y->size();
+  mask->Resize(output_size);
+
+  MaxPoolForward<T><<<CAFFE_GET_BLOCKS(output_size),
+                      CAFFE_CUDA_NUM_THREADS,
+                      0,
+                      context_.cuda_stream()>>>(
+      output_size, X.data<T>(), X.dim32(0), X.dim32(1), X.dim32(2), X.dim32(3),
+      Y->dim32(2), Y->dim32(3), kernel_h(), kernel_w(), stride_h(), stride_w(),
+      pad_t(), pad_l(), Y->mutable_data<T>(), mask->mutable_data<int>());
+  return true;
+
+}
+
+bool MaxPoolWithIndexOp::RunOnDevice() {
+  auto& X = Input(0);
+
+  if (X.IsType<float>()) {
+    return DoRunWithType<float>();
+  } else if (X.IsType<float16>()) {
+    return DoRunWithType<float16>();
+  } else {
+    CAFFE_THROW("Unsupported input type");
+  }
+  return false;
+}
+
+template <typename T>
+bool MaxPoolWithIndexGradientOp::DoRunWithType() {
+  auto& X  = Input(0);
+  auto& dY = Input(1);
+  auto& mask = Input(2);
+  auto* dX = Output(0);
+
+  dX->ResizeLike(X);
+  ConvPoolOpBase<CUDAContext>::ComputePads(vector<int>{X.dim32(2), X.dim32(3)});
+
+  MaxPoolBackward<T><<<CAFFE_GET_BLOCKS(X.size()),
+                       CAFFE_CUDA_NUM_THREADS,
+                       0,
+                       context_.cuda_stream()>>>(
+      X.size(), dY.data<T>(), mask.data<int>(), X.dim32(0), X.dim32(1), X.dim32(2), X.dim32(3),
+      dY.dim32(2), dY.dim32(3), kernel_h(), kernel_w(), stride_h(), stride_w(),
+      pad_t(), pad_l(), dX->template mutable_data<T>());
+  return true;
+}
+
+bool MaxPoolWithIndexGradientOp::RunOnDevice() {
+  auto& X = Input(0);
```

Comment:
Simon, since now the pooling op supports arbitrary dimensions but this is still just 2d, would you mind adding an enforce here to make sure that X is 4-dimensional?